apollo-router/src/router.rs

```diff
@@ -952,6 +958,86 @@ fn generate_event_stream(
     stream
 }
 
+#[cfg(test)]
+struct TestRouterHttpServer {
+    router_http_server: RouterHttpServer,
+    event_sender: mpsc::UnboundedSender<Event>,
+    state_machine_update_notifier: Arc<Notify>,
+}
+
+#[cfg(test)]
+impl TestRouterHttpServer {
+    fn new() -> Self {
+        let (event_sender, event_receiver) = mpsc::unbounded();
+        let state_machine_update_notifier = Arc::new(Notify::new());
+
+        let server_factory = AxumHttpServerFactory::new();
+        let router_factory: OrbiterRouterSuperServiceFactory<YamlRouterFactory> =
+            OrbiterRouterSuperServiceFactory::new(YamlRouterFactory::default());
+        let state_machine = StateMachine::for_tests(
+            server_factory,
+            router_factory,
+            Arc::clone(&state_machine_update_notifier),
+        );
+
+        let listen_addresses = state_machine.listen_addresses.clone();
+        let result = spawn(
+            async move { state_machine.process_events(event_receiver).await }
+                .with_current_subscriber(),
+        )
+        .map(|r| match r {
+            Ok(Ok(ok)) => Ok(ok),
+            Ok(Err(err)) => Err(err),
+            Err(err) => {
+                tracing::error!("{}", err);
+                Err(ApolloRouterError::StartupError)
+            }
+        })
+        .with_current_subscriber()
+        .boxed();
+
+        TestRouterHttpServer {
+            router_http_server: RouterHttpServer {
+                result,
+                shutdown_sender: None,
+                listen_addresses,
+            },
+            event_sender,
+            state_machine_update_notifier,
+        }
+    }
+
+    async fn request(
+        &self,
+        request: crate::graphql::Request,
+    ) -> Result<crate::graphql::Response, crate::error::FetchError> {
+        Ok(reqwest::Client::new()
+            .post(format!("{}/", self.listen_address().await.unwrap()))
+            .json(&request)
+            .send()
+            .await
+            .expect("couldn't send request")
+            .json()
+            .await
+            .expect("couldn't deserialize into json"))
+    }
+
+    async fn listen_address(&self) -> Option<ListenAddr> {
+        self.router_http_server.listen_address().await
+    }
+
+    async fn send_event(&mut self, event: Event) -> Result<(), SendError> {
+        let result = self.event_sender.send(event).await;
+        self.state_machine_update_notifier.notified().await;
+        result
+    }
+
+    async fn shutdown(mut self) -> Result<(), ApolloRouterError> {
```

Comment:
```suggestion
    #[track_caller]
    async fn shutdown(mut self) -> Result<(), ApolloRouterError> {
```

with `track_caller`, it will indicate directly the caller as the cause of the panic from the unwrap

Reply:
```
error: `#[track_caller]` on async functions is a no-op
    --> apollo-router/src/router.rs:1035:5
     |
1035 |       #[track_caller]
     |       ^^^^^^^^^^^^^^^
1036 | /     async fn shutdown(mut self) -> Result<(), ApolloRouterError> {
1037 | |         self.send_event(Event::Shutdown).await.unwrap();
1038 | |         self.router_http_server.shutdown().await
1039 | |     }
     | |_____- this function will not propagate the caller location
```

:oh_no: